File: native/src/parser.rs
use crate::{lex, LexError, Token, TokenKind};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Program {
    pub functions: Vec<Function>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Function {
    pub name: String,
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Statement {
    Call(Call),
    Let { name: String, value: Expression },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Call {
    pub name: String,
    pub arguments: Vec<Expression>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Expression {
    String(String),
    Integer(i64),
    Boolean(bool),
    Variable(String),
    Binary {
        left: Box<Expression>,
        operator: BinaryOperator,
        right: Box<Expression>,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BinaryOperator {
    Add,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParseError {
    pub message: String,
    pub line: usize,
    pub column: usize,
}

pub fn parse(source: &str) -> Result<Program, ParseError> {
    let tokens = lex(source).map_err(ParseError::from)?;
    Parser {
        tokens,
        position: 0,
    }
    .parse()
}

struct Parser {
    tokens: Vec<Token>,
    position: usize,
}

impl Parser {
    fn parse(mut self) -> Result<Program, ParseError> {
        let mut functions = Vec::new();
        while !self.check(TokenKind::Eof) {
            functions.push(self.function()?);
        }
        Ok(Program { functions })
    }

    fn function(&mut self) -> Result<Function, ParseError> {
        self.consume(TokenKind::Fn, "expected 'fn'")?;
        let name = self
            .consume(TokenKind::Identifier, "expected function name")?
            .lexeme;
        self.consume(TokenKind::LParen, "expected '('")?;
        self.consume(TokenKind::RParen, "expected ')'")?;
        self.consume(TokenKind::LBrace, "expected '{'")?;
        let mut body = Vec::new();
        while !self.check(TokenKind::RBrace) {
            if self.check(TokenKind::Eof) {
                return Err(self.error("expected '}'"));
            }
            body.push(self.statement()?);
            if self.check(TokenKind::Semicolon) {
                self.position += 1;
            }
        }
        self.consume(TokenKind::RBrace, "expected '}'")?;
        Ok(Function { name, body })
    }

    fn statement(&mut self) -> Result<Statement, ParseError> {
        if self.check(TokenKind::Let) {
            self.position += 1;
            let name = self
                .consume(TokenKind::Identifier, "expected variable name")?
                .lexeme;
            self.consume(TokenKind::Equal, "expected '='")?;
            return Ok(Statement::Let {
                name,
                value: self.expression()?,
            });
        }
        self.call()
    }

    fn call(&mut self) -> Result<Statement, ParseError> {
        let name = self
            .consume(TokenKind::Identifier, "expected call name")?
            .lexeme;
        self.consume(TokenKind::LParen, "expected '('")?;
        let mut arguments = Vec::new();
        if !self.check(TokenKind::RParen) {
            arguments.push(self.expression()?);
        }
        self.consume(TokenKind::RParen, "expected ')'")?;
        Ok(Statement::Call(Call { name, arguments }))
    }

    fn expression(&mut self) -> Result<Expression, ParseError> {
        let mut expression = self.primary()?;
        while self.check(TokenKind::Plus) {
            self.position += 1;
            expression = Expression::Binary {
                left: Box::new(expression),
                operator: BinaryOperator::Add,
                right: Box::new(self.primary()?),
            };
        }
        Ok(expression)
    }

    fn primary(&mut self) -> Result<Expression, ParseError> {
        let token = self.current();
        match token.kind {
            TokenKind::String => Ok(Expression::String(
                token.lexeme[1..token.lexeme.len() - 1].to_owned(),
            )),
            TokenKind::Integer => Ok(Expression::Integer(
                token.lexeme.parse().expect("lexer emitted integer"),
            )),
            TokenKind::True => Ok(Expression::Boolean(true)),
            TokenKind::False => Ok(Expression::Boolean(false)),
            TokenKind::Identifier => Ok(Expression::Variable(token.lexeme.clone())),
            _ => Err(self.error("expected literal")),
        }
        .inspect(|_| self.position += 1)
    }

    fn consume(&mut self, kind: TokenKind, message: &str) -> Result<Token, ParseError> {
        if self.check(kind) {
            let token = self.current().clone();
            self.position += 1;
            Ok(token)
        } else {
            Err(self.error(message))
        }
    }

    fn check(&self, kind: TokenKind) -> bool {
        self.current().kind == kind
    }

    fn current(&self) -> &Token {
        &self.tokens[self.position]
    }

    fn error(&self, message: &str) -> ParseError {
        ParseError {
            message: message.to_owned(),
            line: self.current().line,
            column: self.current().column,
        }
    }
}

impl From<LexError> for ParseError {
    fn from(error: LexError) -> Self {
        Self {
            message: error.message,
            line: error.line,
            column: error.column,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_minimal_program() {
        let program = parse("fn main() { print(\"Hello AXIOM\") }").unwrap();
        assert_eq!(program.functions[0].name, "main");
        assert_eq!(program.functions[0].body.len(), 1);
    }

    #[test]
    fn reports_missing_body() {
        let error = parse("fn main()").unwrap_err();
        assert_eq!(error.message, "expected '{'");
        assert_eq!(error.line, 1);
    }
}

File: src/axiom/parser.py
"""Recursive-descent parser for the first AXIOM syntax slice."""

from __future__ import annotations

from .ast import ArrayLiteral, Assign, Binary, BooleanLiteral, Break, Call, Continue, FieldAccess, FloatLiteral, For, Function, If, Index, IntegerLiteral, Let, Parameter, Program, Return, StringLiteral, StructDefinition, StructLiteral, Unary, Variable, While
from .lexer import Token, TokenKind, lex


class ParseError(ValueError):
    """Raised when tokens do not match the AXIOM grammar."""


class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.position = 0

    def parse(self) -> Program:
        functions = []
        structs = []
        while not self._check(TokenKind.EOF):
            if self._check(TokenKind.STRUCT):
                structs.append(self._struct())
            else:
                functions.append(self._function())
        return Program(functions, structs)

    def _struct(self) -> StructDefinition:
        self._consume(TokenKind.STRUCT, "expected 'struct'")
        name = self._consume(TokenKind.IDENTIFIER, "expected struct name").lexeme
        self._consume(TokenKind.LBRACE, "expected '{'")
        fields = []
        while not self._check(TokenKind.RBRACE):
            field_name = self._consume(TokenKind.IDENTIFIER, "expected field name").lexeme
            self._consume(TokenKind.COLON, "expected ':'")
            fields.append(Parameter(field_name, self._type_name()))
            if self._check(TokenKind.COMMA) or self._check(TokenKind.SEMICOLON):
                self.position += 1
        self._consume(TokenKind.RBRACE, "expected '}'")
        return StructDefinition(name, fields)

    def _function(self) -> Function:
        self._consume(TokenKind.FN, "expected 'fn'")
        name = self._consume(TokenKind.IDENTIFIER, "expected function name").lexeme
        self._consume(TokenKind.LPAREN, "expected '('")
        parameters = []
        if not self._check(TokenKind.RPAREN):
            parameters.append(self._parameter())
            while self._check(TokenKind.COMMA):
                self.position += 1
                parameters.append(self._parameter())
        self._consume(TokenKind.RPAREN, "expected ')'")
        return_type = None
        if self._check(TokenKind.ARROW):
            self.position += 1
            return_type = self._type_name()
        body = self._block()
        return Function(name, body, parameters, return_type)

    def _block(self):
        self._consume(TokenKind.LBRACE, "expected '{'")
        body = []
        while not self._check(TokenKind.RBRACE):
            if self._check(TokenKind.EOF):
                self._error(self._current(), "expected '}'")
            body.append(self._statement())
            if self._check(TokenKind.SEMICOLON):
                self.position += 1
        self._consume(TokenKind.RBRACE, "expected '}'")
        return body

    def _parameter(self) -> Parameter:
        name = self._consume(TokenKind.IDENTIFIER, "expected parameter name").lexeme
        self._consume(TokenKind.COLON, "expected ':'")
        type_name = self._type_name()
        return Parameter(name, type_name)

    def _type_name(self) -> str:
        type_name = self._consume(TokenKind.IDENTIFIER, "expected type name").lexeme
        while self._check(TokenKind.LBRACKET):
            self.position += 1
            self._consume(TokenKind.RBRACKET, "expected ']' in array type")
            type_name += "[]"
        return type_name

    def _statement(self) -> Call | Let:
        if self._check(TokenKind.LET):
            self.position += 1
            name = self._consume(TokenKind.IDENTIFIER, "expected variable name").lexeme
            type_name = None
            if self._check(TokenKind.COLON):
                self.position += 1
                type_name = self._type_name()
            self._consume(TokenKind.EQUAL, "expected '='")
            return Let(name, type_name, self._expression())
        if self._check(TokenKind.RETURN):
            self.position += 1
            return Return(self._expression())
        if self._check(TokenKind.IF):
            self.position += 1
            condition = self._expression()
            then_body = self._block()
            else_body = []
            if self._check(TokenKind.ELSE):
                self.position += 1
                if self._check(TokenKind.IF):
                    else_body = [self._statement()]
                else:
                    else_body = self._block()
            return If(condition, then_body, else_body)
        if self._check(TokenKind.WHILE):
            self.position += 1
            return While(self._expression(), self._block())
        if self._check(TokenKind.BREAK):
            self.position += 1
            return Break()
        if self._check(TokenKind.CONTINUE):
            self.position += 1
            return Continue()
        if self._check(TokenKind.FOR):
            self.position += 1
            self._consume(TokenKind.LPAREN, "expected '('")
            initializer = None
            if not self._check(TokenKind.SEMICOLON):
                checkpoint = self.position
                if self._check(TokenKind.LET):
                    self.position += 1
                    name = self._consume(TokenKind.IDENTIFIER, "expected variable name").lexeme
                    type_name = None
                    if self._check(TokenKind.COLON):
                        self.position += 1
                        type_name = self._type_name()
                    self._consume(TokenKind.EQUAL, "expected '='")
                    initializer = Let(name, type_name, self._expression())
                else:
                    target = self._primary()
                    if not self._check(TokenKind.EQUAL):
                        self.position = checkpoint
                        self._error(self._current(), "expected loop initializer")
                    self.position += 1
                    initializer = Assign(target, self._expression())
            self._consume(TokenKind.SEMICOLON, "expected ';'")
            condition = self._expression()
            self._consume(TokenKind.SEMICOLON, "expected ';'")
            update = None
            if not self._check(TokenKind.RPAREN):
                checkpoint = self.position
                target = self._primary()
                if not self._check(TokenKind.EQUAL):
                    self.position = checkpoint
                    self._error(self._current(), "expected loop update")
                self.position += 1
                update = Assign(target, self._expression())
            self._consume(TokenKind.RPAREN, "expected ')' ")
            return For(initializer, condition, update, self._block())
        if self._check(TokenKind.IDENTIFIER):
            checkpoint = self.position
            target = self._primary()
            if self._check(TokenKind.EQUAL):
                self.position += 1
                return Assign(target, self._expression())
            self.position = checkpoint
        expression = self._expression()
        if not isinstance(expression, Call):
            self._error(self._current(), "expected statement")
        return expression

    def _call(self) -> Call:
        name = self._consume(TokenKind.IDENTIFIER, "expected call name").lexeme
        self._consume(TokenKind.LPAREN, "expected '('")
        arguments = []
        if not self._check(TokenKind.RPAREN):
            arguments.append(self._expression())
            while self._check(TokenKind.COMMA):
                self.position += 1
                arguments.append(self._expression())
        self._consume(TokenKind.RPAREN, "expected ')'")
        return Call(name, arguments)

    def _string(self) -> StringLiteral:
        token = self._consume(TokenKind.STRING, "expected string literal")
        return StringLiteral(token.lexeme[1:-1])

    def _expression(self):
        return self._or()

    def _or(self):
        expression = self._and()
        while self._check(TokenKind.OR):
            self.position += 1
            expression = Binary(expression, "||", self._and())
        return expression

    def _and(self):
        expression = self._equality()
        while self._check(TokenKind.AND):
            self.position += 1
            expression = Binary(expression, "&&", self._equality())
        return expression

    def _equality(self):
        expression = self._comparison()
        while self._check(TokenKind.EQUAL_EQUAL) or self._check(TokenKind.NOT_EQUAL):
            operator = self._current().lexeme
            self.position += 1
            expression = Binary(expression, operator, self._comparison())
        return expression

    def _comparison(self):
        expression = self._term()
        while (
            self._check(TokenKind.GREATER)
            or self._check(TokenKind.GREATER_EQUAL)
            or self._check(TokenKind.LESS)
            or self._check(TokenKind.LESS_EQUAL)
        ):
            operator = self._current().lexeme
            self.position += 1
            expression = Binary(expression, operator, self._term())
        return expression

    def _term(self):
        expression = self._factor()
        while self._check(TokenKind.PLUS) or self._check(TokenKind.MINUS):
            operator = self._current().lexeme
            self.position += 1
            expression = Binary(expression, operator, self._factor())
        return expression

    def _factor(self):
        expression = self._unary()
        while self._check(TokenKind.STAR) or self._check(TokenKind.SLASH) or self._check(TokenKind.PERCENT):
            operator = self._current().lexeme
            self.position += 1
            expression = Binary(expression, operator, self._unary())
        return expression

    def _unary(self):
        if self._check(TokenKind.BANG) or self._check(TokenKind.MINUS):
            operator = self._current().lexeme
            self.position += 1
            return Unary(operator, self._unary())
        return self._primary()

    def _primary(self):
        token = self._current()
        if token.kind == TokenKind.LPAREN:
            self.position += 1
            expression = self._expression()
            self._consume(TokenKind.RPAREN, "expected ')'")
            return expression
        if token.kind == TokenKind.STRING:
            return self._string()
        if token.kind == TokenKind.INTEGER:
            self.position += 1
            return IntegerLiteral(int(token.lexeme))
        if token.kind == TokenKind.FLOAT:
            self.position += 1
            return FloatLiteral(float(token.lexeme))
        if token.kind in (TokenKind.TRUE, TokenKind.FALSE):
            self.position += 1
            return BooleanLiteral(token.kind == TokenKind.TRUE)
        if token.kind == TokenKind.IDENTIFIER:
            self.position += 1
            if self._check(TokenKind.LBRACE):
                self.position += 1
                fields = []
                while not self._check(TokenKind.RBRACE):
                    field_name = self._consume(TokenKind.IDENTIFIER, "expected field name").lexeme
                    self._consume(TokenKind.COLON, "expected ':'")
                    fields.append((field_name, self._expression()))
                    if self._check(TokenKind.COMMA) or self._check(TokenKind.SEMICOLON):
                        self.position += 1
                self._consume(TokenKind.RBRACE, "expected '}'")
                expression = StructLiteral(token.lexeme, fields)
            elif self._check(TokenKind.LPAREN):
                expression = self._finish_call(token.lexeme)
            else:
                expression = Variable(token.lexeme)
        elif token.kind == TokenKind.LBRACKET:
            self.position += 1
            elements = []
            if not self._check(TokenKind.RBRACKET):
                elements.append(self._expression())
                while self._check(TokenKind.COMMA):
                    self.position += 1
                    elements.append(self._expression())
            self._consume(TokenKind.RBRACKET, "expected ']'")
            expression = ArrayLiteral(elements)
        else:
            self._error(token, "expected expression")
        while self._check(TokenKind.LBRACKET):
            self.position += 1
            index = self._expression()
            self._consume(TokenKind.RBRACKET, "expected ']'")
            expression = Index(expression, index)
        while self._check(TokenKind.DOT):
            self.position += 1
            field = self._consume(TokenKind.IDENTIFIER, "expected field name").lexeme
            expression = FieldAccess(expression, field)
        return expression

    def _finish_call(self, name: str) -> Call:
        self._consume(TokenKind.LPAREN, "expected '('")
        arguments = []
        if not self._check(TokenKind.RPAREN):
            arguments.append(self._expression())
            while self._check(TokenKind.COMMA):
                self.position += 1
                arguments.append(self._expression())
        self._consume(TokenKind.RPAREN, "expected ')'")
        return Call(name, arguments)

    def _consume(self, kind: TokenKind, message: str) -> Token:
        if self._check(kind):
            token = self._current()
            self.position += 1
            return token
        self._error(self._current(), message)

    def _check(self, kind: TokenKind) -> bool:
        return self._current().kind == kind

    def _current(self) -> Token:
        return self.tokens[self.position]

    def _peek(self) -> Token:
        return self.tokens[min(self.position + 1, len(self.tokens) - 1)]

    @staticmethod
    def _error(token: Token, message: str) -> None:
        raise ParseError(f"{message} at {token.line}:{token.column}")


def parse(source: str) -> Program:
    return Parser(lex(source)).parse()